BDArmory/Control/MissileFire.cs

```diff
@@ -3074,14 +3078,18 @@ private void UpdateSelectedWeaponState()
             if (selectedWeapon != null && (selectedWeapon.GetWeaponClass() == WeaponClasses.Bomb || selectedWeapon.GetWeaponClass() == WeaponClasses.Missile || selectedWeapon.GetWeaponClass() == WeaponClasses.SLW))
             {
                 //Debug.Log("[BDArmory.MissileFire]: =====selected weapon: " + selectedWeapon.GetPart().name);
-                if (!CurrentMissile || CurrentMissile.GetPartName() != selectedWeapon.GetPartName())
+                if (!CurrentMissile || CurrentMissile.GetPartName() != selectedWeapon.GetPartName() || CurrentMissile.engageRangeMax != float.Parse(selectedWeaponString.Substring(selectedWeaponString.IndexOf(";") + 1)))
```

Comment:
The modifying the selectedWeaponName came out of needing to deal with UpdateList() as well as selectedWeapon, to allow missiles with different ranges to appear in the WM Weapon list/be manually selectable, while also needing to distinguish variable ranges with just a weaponIndex int, and the shortname method was cleaner than the original method I was looking at for modifying the UpdateList logic.
I do agree that it's suboptimal, so I'll take another look at alternative methods.

Reply:
Another reason for not basing this on special characters in the `shortName` is that it'll break if anyone makes a weapon that includes that character in its `shortName`.

From what I can see of the current implementation, this should be relatively easy to change to avoid the problematic string manipulation by including the `selectedWeaponsEngageRangeMax` stuff above and adding a `GetShortNameBase()` to `IBDWeapon.cs` that returns the short name without the added range (regular weapons would just return `GetShortName()`), which you'd just store before modifying `shortName` and then the `GetShortName().Substring(0, GetShortName().IndexOf(";"))` calls would be replaced with `GetShortNameBase()` and `float.TryParse(selectedWeaponString.Substring(selectedWeaponString.IndexOf(";") + 1), out float launchRange); mslRange = launchRange;` would be replaced with `mslRange = selectedWeaponsEngageRangeMax`.

Also, I personally prefer `$"{OriginalShortName} ({engageRangeMax}m)"` instead of `$"{OriginalShortname};{engageRangeMax}"`, btw.

Reply:
i've reverted the method that modifies the shortname and gotten something similar to your original suggestion implemented